tests/pass/float.rs

```diff
@@ -148,155 +260,250 @@ fn basic() {
     assert_eq!(34.2f64.abs(), 34.2f64);
 }
 
+/// Test casts from floats to ints and back
+macro_rules! test_ftoi_itof {
+    (
+        f: $fty:ty,
+        i: $ity:ty,
+        // Int min and max as float literals
+        imin_f: $imin_f:literal,
+        imax_f: $imax_f:literal $(,)?
+    ) => {{
+        /// By default we test float to int `as` casting as well as to_int_unchecked
+        fn assert_ftoi(f: $fty, i: $ity, msg: &str) {
+            test_both_cast::<$fty, $ity>(f, i, msg);
+        }
+
+        /// Unrepresentable values only get tested with `as` casting, not unchecked
+        fn assert_ftoi_unrep(f: $fty, i: $ity, msg: &str) {
+            let msg = format!("{} -> {}: {msg}", stringify!($fty), stringify!($ity));
+            assert_eq_msg::<$ity>(f as $ity, i, &msg);
+        }
+
+        /// Int to float checks
+        fn assert_itof(i: $ity, f: $fty, msg: &str) {
+            let itof_msg = format!("{} -> {}: {msg}", stringify!($ity), stringify!($fty));
+            assert_eq_msg::<$fty>(i as $fty, f, &itof_msg);
+        }
+
+        /// Check both float to int and int to float
+        fn assert_bidir(f: $fty, i: $ity, msg: &str) {
+            assert_ftoi(f, i, msg);
+            assert_itof(i, f, msg);
+        }
+
+        /// Check both float to int and int to float
+        fn assert_bidir_unrep(f: $fty, i: $ity, msg: &str) {
+            assert_ftoi_unrep(f, i, msg);
+            assert_itof(i, f, msg);
+        }
+
+        let fbits = <$fty>::BITS;
+        let ibits = <$ity>::BITS;
+        let imax = <$ity>::MAX;
+        let imin = <$ity>::MIN;
+        let isigned = <$ity>::SIGNED;
+
+        #[allow(overflowing_literals)]
+        let imin_f = $imin_f;
+        #[allow(overflowing_literals)]
+        let imax_f = $imax_f;
+
+        // Skip unchecked cast if negative numbers are unrepresentable
+        let assert_ftoi_neg = if isigned { assert_ftoi } else { assert_ftoi_unrep };
+        // Skip unchecked cast when int min/max would be unrepresentable
+        let assert_ftoi_big = if ibits < fbits { assert_ftoi } else { assert_ftoi_unrep };
```

Comment:
This is not a strict rule, is it? A hypothetical f33 type could probably not represent all u32 even though it is "bigger", since bits are lost to exponent and sign?
I'm okay with having a heuristic here, but there should be comments explaining that.

Reply:
I think the actual rule is `F::SIGNIFICAND_BITS + 1 >= I::BITS` to have a guaranteed representation for all values, since the entire integer can be contained without truncation. I'll just change this since it may come up for something like bf16 if that winds up being added.

Reply:
Why `+ 1`?